keps/sig-api-machinery/2019-02-13-custom-resource-definitions.md

```diff
@@ -0,0 +1,212 @@
+reviewers:
+  - "@bgrant0607"
+  - "@smarterclayton"
+  - "@lavalamp"
+  - "@deads2k"
+approvers:
+  - "@lavalamp"
+  - "@smarterclayton"
+  - TBD
+editor: TBD
+creation-date: "2013-02-13"
+last-updated: "2013-02-13"
+status: implemented
+---
+
+# Custom Resource Definitions
+
+## Summary
+
+This KEP covers the addition and usage of Custom Resource Definitions. This KEP is different than most KEPs
+because it is an attempt to retrospectively describe the development of a feature in Kubernetes that 
+already exists. We are doing this because custom resource definitions were created and evolved before the
+KEP process was put in place. Because of this, there is no single place that describes the reasons for
+custom resource definitions and how they should be used in Kubernetes.
+
+## Motivation
+
+The primary motivation for custom resources was to provide a simple easy to use extension mechanism
+for Kubernetes clusters to enable customization by Kubernetes users and to facilitate a rich
+ecosystem of plugins for Kubernetes clusters. If Kubernetes is the kernel, CustomResourceDefinitions
+are intended to be loadable kernel modules.
+
+### Goals
+
+   * Provide an easy to use mechanism for the dynamic addition of new API types to Kubernetes.
+   * Enable API extensibility without significant additional load on operators of Kubernetes clusters.
+   * Enable an ecosystem of plugin providers who can produce and distribute extensions that add value to end-user clusters.
+
+### Non-Goals
+   * Replace all existing built-in API objects with Custom Resources.
+   * Enable every possible extension API. Unsupported patterns should use aggregated API servers.
+   * Develop an arbitrary key-value store for applications. Custom resources are control-plane objects.
+   * Develop tooling for managing the installation ordering or dependencies of custom resources. This is deferred to other KEPs.
+   * Expressing or managing inheritance, referential integrity, or other inter-object relationships.
+   * Document how to create a custom resource definition. This is done [elsewhere](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/)
+
+
+## Implementation History
+
+Third-party resources were originally [proposed](https://github.com/kubernetes/kubernetes/pull/11781) 
+shortly after Kubernetes 1.0. At the time it was clear that making the Kubernetes API server
+extensible was critical to the long-term health of the project and there were two different
+approaches for extension proposed:
+   * Plugin resources (initial "ThirdPartyResource" eventually "CustomResourceDefinitions")
+   * Delgation to an "aggregated API server"
+
+ThirdPartyResources were added to the API server, but the API server was actually really not ready
+for them, much of the dynamic discovery and grouping mechanisms were primitive, or non-existent.
+As a result, and because there was little initial support from the core API machinery maintainers
```

Comment:
I think some people would tell the story a little differently :)

No matter how it's told, this isn't a good sort of story to have in this document.

I would just say TPRs were an inexpensive experiment, finished just enough to be useful, and we learned a lot from them.

Reply:
> I would just say TPRs were an inexpensive experiment, finished just enough to be useful, and we learned a lot from them.

Agree.